### DISCUSSION TITLE
DCD19 - HASIL VISUALISASI HEATMAP DI SHINY  BERBEDA DENGAN HASIL DI R

### DISCUSSION THREAD

**Discussion by Micoalgo** [teamalgoritma/community#673] [Q&A] [answered]
Mohon Solusinya, Apakah yg menyebabkan adanya perbedaan hasil visualisasi Heatmap di Shiny berbeda dg hasil visualisasi di R ? Di Shiny Heatmap tdk tersebar dan hanya tersentralisasi di jam tertentu saja, sedangkan keduanya berangkat dari kode yg sama, yaitu sbb ;

1. Visualisasi di Shiny
![image](https://github.com/teamalgoritma/community/assets/123436963/0cc8e6dd-8e31-40da-a5c6-6038f14e5e9c)

2. Visualisasi di R
![image](https://github.com/teamalgoritma/community/assets/123436963/3ae38485-f286-4a04-b47f-44d0a7075f51)

3. Kode
![image](https://github.com/teamalgoritma/community/assets/123436963/cb8b9c36-ccce-4740-bd9e-914febd05422)
![image](https://github.com/teamalgoritma/community/assets/123436963/8f4f4dc8-70e9-480c-ab2e-d2f380530d82)
![image](https://github.com/teamalgoritma/community/assets/123436963/fa306cf2-433f-4ae1-839a-19fcb3ff90da)
![image](https://github.com/teamalgoritma/community/assets/123436963/8bca35e6-1035-42e7-921e-b3803b70aa6e)

Terimakasih

**Comment by jafarhf1** [comment]
Dear Pak @Micoalgo,

Mengenai pertanyaannya, boleh tolong dilampirkan juga kode pada bagian ui.R dan server.R yang terhubung dengan plot heatmap tersebut?

Terima kasih

**Reply by Micoalgo** [reply]
ui.R
<img width="967" alt="image" src="https://github.com/teamalgoritma/community/assets/123436963/c29ceb82-97b4-4671-af97-546849c359e6">

**Reply by Micoalgo** [reply]
Server.R
<img width="996" alt="image" src="https://github.com/teamalgoritma/community/assets/123436963/78bcdfff-cb42-4c6d-9633-e4e3f55540b4">

<img width="1014" alt="image" src="https://github.com/teamalgoritma/community/assets/123436963/330ee164-dbf3-413c-9f5c-1ca95b1e5937">

<img width="1011" alt="image" src="https://github.com/teamalgoritma/community/assets/123436963/279a6bec-2b44-494b-a446-a2b9202ef1f7">

<img width="964" alt="image" src="https://github.com/teamalgoritma/community/assets/123436963/d2b1d016-1e3f-4f2e-bb85-ef94e66c4b3a">

<img width="965" alt="image" src="https://github.com/teamalgoritma/community/assets/123436963/ddfce1c0-58b7-41af-af2a-ec2d71fb65ed">

**Reply by Micoalgo** [reply]
Maaf ada yg tertinggal :
<img width="606" alt="image" src="https://github.com/teamalgoritma/community/assets/123436963/5511fe4d-c02b-4412-8e93-339bf180c048">

**Reply by divakartika** [reply]
Dear Pak @Micoalgo ,

Dapat dicoba jika plot pada Shiny server tidak menggunakan `select_input23` sehingga `filter(unit == "ANAK")` apakah menghasilkan plot yang sesuai?

**Comment by Micoalgo** [comment]
Sudah saya coba juga semalam mas... Hasilnya masih berbeda juga antara
Shiny dan R

**Reply by divakartika** [reply]
Dear Pak @Micoalgo ,

Apakah kesalahan ini hanya terjadi pada unit `ANAK`? Selain itu dapat diinfokan apakah data `bpjs_kpad2_heatmap` terdapat di global.R? Karena data tersebut digunakan dalam `selectInput`.

**Reply by Micoalgo** [reply]
Kesalahan tidak hanya di menu Anak Mba Diva, tapi di unit lain juga demikian , misal di IGD ;

Shiny 
<img width="707" alt="image" src="https://github.com/teamalgoritma/community/assets/123436963/01084d05-07e0-4f66-b283-175c6e185967">


R
<img width="696" alt="image" src="https://github.com/teamalgoritma/community/assets/123436963/51a0e767-8683-485d-995d-3aa127692a72">

**Reply by Micoalgo** [reply]
bpjs_kpad2_heatmap ada pada Global.R, Berikut Tampilan utk Global.R ;

<img width="951" alt="image" src="https://github.com/teamalgoritma/community/assets/123436963/d11f97e5-8fb5-4791-b64e-591badc8befb">

<img width="957" alt="image" src="https://github.com/teamalgoritma/community/assets/123436963/bb134c3b-e14c-42b7-bdf1-c598904b9df7">

<img width="950" alt="image" src="https://github.com/teamalgoritma/community/assets/123436963/d49344e7-2479-4ece-975f-d9db5d0e2256">

<img width="485" alt="image" src="https://github.com/teamalgoritma/community/assets/123436963/929d27a1-16f9-4114-bfdf-bffbeb4252a6">

**Reply by Micoalgo** [reply]
Apakah ada Library yg kurang, atau ada masalah yg lain ya ? Terimakasih

(Saya coba tambahkan Library heatmaply dan library shinyHeatmaply tapi belum solved)

**Comment by kevbow** [comment]
Dear Pak @Micoalgo,

Mungkin bisa dicoba tidak menggunakan ggplotly, gunakanlah ggplot biasa karena hal seperti ini biasanya terjadi karena fitur yang dikeluarkan tidak tersedia pada plotly.

**Comment by Micoalgo** [comment]
Saya coba Mas Kevin... Terimakasih 🙏

**Reply by Micoalgo** [reply]
Selamat Siang Mas @kevbow 

Setelah saya coba yang disarankan didaptkan hasil sbb ; 
(Screen Shoot Bagian atas dan bawah kode  di Server dan ui serta hasil visualisasinya di Shiny dan di R)

Server.R
<img width="720" alt="image" src="https://github.com/teamalgoritma/community/assets/123436963/0fbee7a8-146a-4d4a-89df-f9e43f71c1a5">
.....

<img width="731" alt="image" src="https://github.com/teamalgoritma/community/assets/123436963/a3661d73-1569-41de-af45-2c5c58669727">

ui.R
<img width="816" alt="image" src="https://github.com/teamalgoritma/community/assets/123436963/34629e36-0564-4669-ba08-692fc23386b3">

Visualisasinya msh terdapat perbedaan yaitu;

Shiny (plot tidak interaktif)
<img width="727" alt="image" src="https://github.com/teamalgoritma/community/assets/123436963/621cf5f6-db6b-43e7-bef9-163e4193f33e">

R (plot tidak interaktif)
<img width="683" alt="image" src="https://github.com/teamalgoritma/community/assets/123436963/4fbad72b-e7b0-4fa1-94c2-cb1cd74a85d8">

<img width="686" alt="image" src="https://github.com/teamalgoritma/community/assets/123436963/01703b94-fe89-44dd-92e9-b8027957cef8">

**Reply by Micoalgo** [reply]
Masih didapatkan hasil yg belum diharapkan dan plot menjadi tidak interaktif sehingga kurang praktis untuk dibaca, mohon arahannya. Terimakasih

**Reply by jafarhf1** [reply]
Dear Pak @Micoalgo,

Pada file ui.R, boleh dicoba pak untuk ditambahkan parameter `width`  dan `height` pada fungsi `plotlyOutput()` nya sehingga menjadi seperti ini : 
```
plotlyOutput(outputId = "heatmap_try2_bpjs_anak",  height = "100%",  width = "100%")
```

Semoga membantu pak, Terima kasih

**Reply by Micoalgo** [reply]
Sudah  saya coba Mas @jafarhf1 , sepertinya belum ada perubahan.
Berikut kode dan visualisasi nya ;
<img width="785" alt="image" src="https://github.com/teamalgoritma/community/assets/123436963/647a699f-0af7-46a2-84c6-17795c9db65a">

<img width="709" alt="image" src="https://github.com/teamalgoritma/community/assets/123436963/7a68169e-e626-4e85-a600-6257ed64e132">

Terimakasih

**Comment by ramdhinafinita** [comment]
Dear pak @Micoalgo 
Saya coba jawab kendala bapak. Mungkin bisa dicoba beberapa hal berikut
pada ui.R:
pada parameter choices, mungkin bisa dicoba menggunakan levels(bpjs_kpad2_heatmap$unit) 
jika kolom unit sebelumnya sudah di ubah menjadi as.factor. 

line 3372 filter dicoba jangan unique tetapi pakai level.. atau dicoba dulu jangan sesuai select input tp masih yg statis unit == "ANAK" biar bisa di track error di kode atau inputnya.

pada ggplotly di server.R oba dihapus type ="heatmap" nya.

**Reply by Micoalgo** [reply]
Saya coba dulu ya Mba...Terimakasih

**Reply by ramdhinafinita** [reply]
Baik pak, dan bagian server hanya memasukkan langkah aggregasi persiapan data plot heatmap nya saja. Sedangkan di bagian global langkah persiapan data nya. Mungkin bisa dicoba juga pak dihapus salah satunya, karena jika dilihat pada global.R dengan server.R isinya sama.

**Reply by Micoalgo** [reply]
Sudah saya coba Mba @ramdhinafinita , didapatkan hasil dan pesan error sbb ;

di ui.R
<img width="804" alt="image" src="https://github.com/teamalgoritma/community/assets/123436963/a9a57941-453f-4d1d-8253-eef2491a0ffc">

di Server.R
unit di factor kan
<img width="796" alt="image" src="https://github.com/teamalgoritma/community/assets/123436963/b8ada0f9-11cc-45ba-9ce9-a7b7387f2f34">

unit ANAK saja
<img width="762" alt="image" src="https://github.com/teamalgoritma/community/assets/123436963/818bcb61-4fa1-4b9c-8422-91348a507ba6">

error pd Shiny
<img width="725" alt="image" src="https://github.com/teamalgoritma/community/assets/123436963/1a9d4f27-e94e-4bbf-aa4b-458a62825bfc">

**Reply by ramdhinafinita** [reply]
pada baris 3374  jangan gunakan lagi levels(unit=="ANAK") pak.. Jika mau cek, pakainya filter() seperti code sebelumnya.
![image](https://github.com/teamalgoritma/community/assets/64061969/6ca1e09d-35ae-4679-b2d5-1e237e4c15fe)

**Reply by Micoalgo** [reply]
Sudah saya coba lagi Mba, Global terkait data clean saja dan server terkait lanjutannya setelah clean tsb, tapi belum ada perubahan...

**Comment by nadyadtm** [comment]
Baik pak, mungkin untuk bagian heatmap ini bisa dicoba dulu membuat plot dalam bentuk yang lain terlebih dahulu (seperti bar plot contohnya). Jika bisa berarti kemungkinan besar ada yang salah dalam pembuatan heatmapnya. Jika tidak muncul apa-apa juga berarti ada masalah saat filtering. Semoga membantu

**Reply by Micoalgo** [reply]
So far barplot dan selain heatmap tidak ada masalah Mba...hanya di Heatmap saja yg bermasalah ketika di Shiny nya...sementara visualisasi heatmap tsb sepertinya saat ini paling relevan utk menunjukan persebarannya pasien menurut hari dan jam nya. Ketika di deploy di Rpubs Heatmap dpt tampil dg baik...tapi tidak ketika di Shiny...

**Comment by jafarhf1** [comment, marked as answer]
Dear Pak @Micoalgo,

Setelah melihat kembali kode yang Bapak lampirkan, ternyata ada satu baris kode yang menyebabkan hal tersebut, yaitu pada bagian perubahan format tipe data date untuk kolom `Tanggal Tagihan`. Kolom tersebut seharusnya memiliki informasi tanggal beserta dengan waktunya (jam menit detik). Akan tetapi, karena Bapak mengubah formatnya menjadi hanya tanggal saja `"%m/%d/%y"`, maka informasi waktunya menjadi hilang. Hal inilah yang menyebabakan heatmap yang ditampilkan tidak sesuai, karena memang pada heatmap tersebut membutuhkan informasi jam pada sumbu y nya. Oleh karena itu, silakan dicomment atau dihapus untuk satu baris kode tersebut.

Demikian yang dapat kami sampaikan, semoga membantu

**Reply by Micoalgo** [reply]
Setelah di coba, Alhamdulillah sudah terjawab masalahnya. Terimakasih Mas @jafarhf1 , Mba @ramdhinafinita , Mba @nadyadtm , Mba @divakartika , Mas @kevbow dan Algoritma atas bantuan dan arahannnya
